### ISSUE TITLE
deviceptr vs present (OpenACC backend)

### ISSUE THREAD

**Issue by szaghi** [szaghi/FUNDAL#4]
It seems I have not completely understood the meaning (and correct usage) of _deviceprt_ and _present_ directives.

From the OpenACC specs I have found that:

#### deviceptr
> The deviceptr clause may appear on structured data and compute constructs and declare directives.
> The deviceptr clause is used to declare that the pointers in var-list are device pointers, so the
> data need not be allocated or moved between the host and device for this pointer.
>
> In C and C++, the vars in var-list must be pointer variables.
>
> In Fortran, the vars in var-list must be dummy arguments (arrays or scalars), and may not have the
> Fortran pointer, allocatable, or value attributes.
> For data in shared memory, host pointers are the same as device pointers, so this clause has no effect.

#### present
> The present clause may appear on structured data and compute constructs and declare directives. The present clause specifies that vars in var-list are in shared memory or are already present in the current device memory due to data regions or data lifetimes that contain the construct on which the present clause appears.
>
>For each var in var-list, if var is in shared memory, no action is taken; if var is not in shared memory,
the present clause behaves as follows:
• At entry to the region:
>– An attach action is performed if var is a pointer reference, and a present increment
action with the structured reference counter is performed if var is not a null pointer.
• At exit from the region:
– If the structured reference counter for var is zero, no action is taken.
– Otherwise, a detach action is performed if var is a pointer reference, and a present decrement
action with the structured reference counter is performed if var is not a null pointer. If
both structured and dynamic reference counters are zero, a delete action is performed.

Moreover, from [OpenACC best practice](https://github.com/OpenACC/openacc-best-practices-guide) we read

> * `present` - The listed variables are already present on the device, so no
  further action needs to be taken. This is most frequently used when a data
  region exists in a higher-level routine.
> * `deviceptr` - The listed variables use device memory that has been managed
  outside of OpenACC, therefore the variables should be used on the device
  without any address translation. This clause is generally used when OpenACC
  is mixed with another programming model, as will be discussed in the
  interoperability chapter.

To my understanding the current implementation of FUNDAL is wrong: we use `deviceptr` directive in the same scope where variable are defined `pointer`, thus GNU gfortran raises an error and stop the compilation while Nvidia nvfortran does not complain and compile the library successful. If I use `present` instead of `deviceptr` both compilers compile the library, but both of them do not work at runtime (with different errors). 

In summary:

+ Nvidia nvfortran compiles and works with `deviceptr`, compiles and does not work with `present`;
+ GNU gfortran compiles only with `present`, but does not work;

I think we need to open some request help on dedicated OpenACC-related forums.

**Comment by giacrossi** [comment]
Yes, I agree that the only thing to do is to ask on Nvidia forum: you can link this project, put two small code snippets (one for the allocation, the other for the initialization, where deviceptr is used) and ask them about that.

**Comment by szaghi** [comment]
@giacrossi I will create a minimal working example reproducing the issue and I'll post on Nvidia forum

**Comment by szaghi** [comment]
I have just created a post into the Nvidia forum:

https://forums.developer.nvidia.com/t/deviceptr-vs-present-openacc-directives/284954

**Comment by giacrossi** [comment]
Could this issue be closed?

**Comment by szaghi** [comment]
@giacrossi 

Yes, it can be closed although the issue is not resolved: GNU gfortran still requires `present` clause instead of `deviceptr` while NVidia does the contrary and the experts did not address our clarification request. Currently, the cpp macro workaround enables both GNU and NVidia compilers.
